repository/src/fetch.rs

```diff
@@ -0,0 +1,269 @@
+    let mut result = Vec::new();
+    // TODO: making this concurrent causes a god damn lifetime annoying error
+    for b in local_branches {
+        result.push((b.clone(), raw.locate_branch(b).await?))
+    }
+    Ok(result.into_iter().collect())
+}
+
+#[derive(Debug, Error)]
+enum CommitError {
+    #[error("raw repo error: {0}")]
+    Raw(#[from] raw::Error),
+    #[error("failed to parse commit ({1}): {0}")]
+    Commit(anyhow::Error, CommitHash),
+}
+
+async fn read_commits<T: RawRepository>(
+    this: &DistributedRepository<T>,
+    ancestor: CommitHash,
+    descendant: CommitHash,
+) -> Result<Vec<(Commit, CommitHash)>, CommitError> {
+    let commits = this.raw.query_commit_path(ancestor, descendant).await?;
+    let commits = stream::iter(commits.iter().cloned().map(|c| {
+        let raw = &this.raw;
+        async move { raw.read_semantic_commit(c).await.map(|x| (x, c)) }
+    }))
+    .buffered(256)
+    .collect::<Vec<_>>()
+    .await;
+    let commits = commits.into_iter().collect::<Result<Vec<_>, _>>()?;
+    let commits = commits
+        .into_iter()
+        .map(|(commit, hash)| {
+            from_semantic_commit(commit)
+                .map_err(|e| (e, hash))
+                .map(|x| (x, hash))
+        })
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|(e, c)| CommitError::Commit(e, c))?;
+    Ok(commits)
+}
+
+pub async fn fetch<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    _network_config: &NetworkConfig,
+    known_peers: &[Peer],
+) -> Result<(), Error> {
+    add_remotes(this, known_peers).await?;
+    this.raw.fetch_all().await?;
+
+    let remote_branches = this.raw.list_remote_tracking_branches().await?;
+    let finalization_proofs = retrieve_fps(this, &remote_branches).await?;
+    // Retrieve local branches to skip if the remote tracking branch is already in the local repo.
+    let local_branches = retrieve_local_branches(&this.raw).await?;
+
+    let mut last_finalized_commit_hash =
+        this.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+    let mut last_finalization_proof = format::fp_from_semantic_commit(
+        this.raw
+            .read_semantic_commit(this.raw.locate_branch(FP_BRANCH_NAME.into()).await?)
+            .await?,
+    )?;
+    let last_header = this.get_last_finalized_block_header().await?;
+    let reserved_state = this.get_reserved_state().await?;
+
+    let mut next_step_branches = Vec::new();
+
+    // step 1: update finalization
+    'branch_loop: for (remote_name, branch_name, commit_hash) in remote_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
+
+        // Skip if the branch is `fp`.
+        if branch_name == FP_BRANCH_NAME {
+            continue;
+        }
+
+        // Check if the branch is already in the local repo.
+        if local_branches
+            .iter()
+            .any(|(_, commit_hash_)| *commit_hash_ == commit_hash)
```

Comment:
```suggestion
            .any(|(_, local_branch_commit_hash)| *local_branch_commit_hash == remote_branch_commit_hash)
```
it doesn't have to be same as this but I think naming should be fixed here

Reply:
fixed